src/utils/collections.zig

```diff
@@ -62,3 +62,265 @@ pub fn RecyclingList(
+            .{ .after = left - 1 }
+        else if (items[left] < search_term)
+            .{ .after = left }
+        else
+            unreachable;
+    }
+    var index = maybe_index.?;
+
+    // match found, move to edge if there are duplicates
+    switch (which) {
+        .any => {},
+        .first => while (index > 0 and items[index - 1] == search_term) {
+            index -= 1;
+        },
+        .last => while (index < items.len - 1 and items[index + 1] == search_term) {
+            index += 1;
+        },
+    }
+
+    return .{ .found = index };
+}
+
+const expect = std.testing.expect;
+const eql = std.mem.eql;
+
+test SortedSet {
+    var set = SortedSet(u64).init(std.testing.allocator);
+    defer set.deinit();
+
+    // add/contains
+    try expect(!set.contains(3));
+    try set.put(3);
+    try expect(set.contains(3));
+    try set.put(0);
+    try set.put(2);
+    try set.put(1);
+    try set.put(4);
+    try set.put(5);
+
+    // remove
+    try expect(set.remove(5));
+    try expect(!set.contains(5));
+    try expect(!set.remove(5));
+    try set.put(5);
+    try expect(set.contains(5));
+
+    // ordering
+    for (set.items(), 0..) |key, i| {
+        try expect(key == i);
+    }
+}
+
+test "SortedSet range" {
+    var set = SortedSet(u8).init(std.testing.allocator);
+    defer set.deinit();
+
+    try set.put(5);
+    try set.put(3);
+    try set.put(1);
+    try set.put(3);
+
+    try expect(eql(u8, &.{ 1, 3, 5 }, set.range(null, null)));
+    try expect(eql(u8, &.{}, set.range(0, 0)));
+    try expect(eql(u8, &.{}, set.range(10, 10)));
+    try expect(eql(u8, &.{}, set.range(10, 11)));
+    try expect(eql(u8, &.{}, set.range(12, 11)));
+    try expect(eql(u8, &.{1}, set.range(null, 3)));
+    try expect(eql(u8, &.{ 1, 3 }, set.range(null, 4)));
+    try expect(eql(u8, &.{ 1, 3 }, set.range(null, 5)));
+    try expect(eql(u8, &.{ 1, 3, 5 }, set.range(null, 6)));
+    try expect(eql(u8, &.{ 1, 3, 5 }, set.range(0, null)));
+    try expect(eql(u8, &.{ 1, 3, 5 }, set.range(1, null)));
+    try expect(eql(u8, &.{ 3, 5 }, set.range(2, null)));
+    try expect(eql(u8, &.{ 3, 5 }, set.range(3, null)));
+    try expect(eql(u8, &.{5}, set.range(4, null)));
+    try expect(eql(u8, &.{5}, set.range(5, null)));
+    try expect(eql(u8, &.{ 1, 3, 5 }, set.range(1, 6)));
+    try expect(eql(u8, &.{ 1, 3 }, set.range(1, 5)));
+    try expect(eql(u8, &.{ 1, 3 }, set.range(1, 4)));
+    try expect(eql(u8, &.{1}, set.range(1, 3)));
+    try expect(eql(u8, &.{1}, set.range(1, 2)));
+    try expect(eql(u8, &.{}, set.range(1, 1)));
+    try expect(eql(u8, &.{ 3, 5 }, set.range(2, 6)));
+    try expect(eql(u8, &.{ 3, 5 }, set.range(3, 6)));
+    try expect(eql(u8, &.{5}, set.range(4, 6)));
+    try expect(eql(u8, &.{5}, set.range(5, 6)));
+    try expect(eql(u8, &.{3}, set.range(3, 4)));
+    try expect(eql(u8, &.{}, set.range(3, 3)));
+    try expect(eql(u8, &.{}, set.range(2, 3)));
+    try expect(eql(u8, &.{}, set.range(2, 2)));
```

Comment:
Similar comment as in the insert_shreds PR, we get better error messages if we use the more specific `std.testing.expect*` functions, in this case `std.testing.expectEqualSlices`.

Reply:
6a82fd1125d3a723db15420c757c8dde63634711